File: Python/Seccion_2/Metodos_de_Strings.py
"""
    Los métodos de strings son funciones que se utilizan para realizar operaciones comunes en cadenas de texto.
"""

# len(): devuelve la longitud de la cadena de texto.
cadenaDeTexto1 = "lorem ipsum"
print(f"La longitud de la cadena de texto 1 es: { len(cadenaDeTexto1) }")
print(f"La longitud de la cadena de texto 2 es: { len(' lorem ipsum otra cosa ') }")




from os import system
system("cls") # Limpia la consola del texto previamente impreso.

# count(): Cuenta el número de concurrencias de una subcadena en una cadena.

cadenaDeTexto2 = "Esto, ok, es una cadena de texto, ok, que contiene un mensaje, ok."
# cadenaDeTexto2 = "Esto, es una cadena de texto, que contiene un mensaje."

print(f"La oracion es: { cadenaDeTexto2 }")
print(f"La palabra ok aparece: { cadenaDeTexto2.count('ok') } veces")
print("Muy bien, no usaste la palabra: ok" if cadenaDeTexto2.count('ok') == 0 else f"Puedes mejorarlo, usaste la palabra: ok {cadenaDeTexto2.count('ok')} veces")



system("cls")


# upper(): convierte la cadena de texto en mayúsculas.
nombreUsuario = "hAonteR" 

print( f"El nombre de usuario es: {nombreUsuario}")
print(f"El nombre del usuario en mayusculas es: {nombreUsuario.upper()}")



system("cls")



# lower(): convierte la cadena de texto en minúsculas.
nombreUsuarioMayusculas = "GuiLLERMo"

print( f"El nombre de usuario es: {nombreUsuarioMayusculas}")
print(f"El nombre del usuario en minúsculas es: {nombreUsuarioMayusculas.lower()}")



system("cls")



# capitalize(): convierte la primera letra de la cadena de texto en mayúscula.
nombreUsuarioMayusculas = "dOUGLAS"

print( f"El nombre de usuario es: {nombreUsuarioMayusculas}")
print(f"El nombre del usuario en capitalizado es: {nombreUsuarioMayusculas.capitalize()}")




system("cls")



# replace(): reemplaza una subcadena de una cadena de texto.

# Ejemplo 1
oracion = "El usuario se encuentra offline." 

print(f"{oracion}")
print(f"{oracion.replace('offline', 'online')}")
print(f"{oracion.replace('online', 'offline')}")
print(f"{oracion.replace('offline', 'online')}")
print(f"{oracion.replace('offline', 'con problemas de internet')}")



system("cls")



# split(): divide una cadena de texto en una lista(array) de subcadenas utilizando un separador especificado.

texto = "Hola Mundo, estamos aprendiendo a usar Python, en Lexpin, con Diego Rodriguez"
listaDePalabras = texto.split(" ")

print(f"La oracion completa es: {texto}")
print(f"La lista de palabras es: {listaDePalabras} \n") # ['Hola Mundo,', 'estamos', 'aprendiendo', 'a', 'usar', 'Python,', 'en', 'Lexpin,', 'con', 'Diego', 'Rodriguez']


# Podemos tambien acceder a los indices de manera individual.
print("="*25 + " LISTA DE PALABRAS " + "="*25)
print(listaDePalabras[0])
print(listaDePalabras[1])
print(listaDePalabras[2])
print(listaDePalabras[3])
print(listaDePalabras[4])
print(listaDePalabras[5])
print(listaDePalabras[6])
print(listaDePalabras[7])
print(listaDePalabras[8])
print(listaDePalabras[9])
print(listaDePalabras[10])
print(listaDePalabras[11])
print("="*69)





system("cls")





# strip(): elimina los caracteres especificados del comienzo y del final de la cadena de texto.
texto = "---Hola Mundo-----------"
print(f"La cadena original es: {texto}") #  La cadena original es: ---Hola Mundo-----------
print(f"Ahora la cadena es: {texto.strip('-')}\n\n") # Ahora la cadena es: Hola Mundo

# strip es parecido al metodo trim() en JavaScript solo que trim() elimina espacios exclusivamente.


system("cls")






# join(): Une los elementos de una lista(array) en una cadena separados por un carácter específico.

print("Separamos la cadena donde exista un espacio.")
texto = "Hola Lexpin"
arrayDePalabras = texto.strip('-').split(' ')
print(f"La lista de palabras es: {arrayDePalabras}") #La lista de palabras es: ['Hola', 'Lexpin']
print(f"El indice 0 es: {arrayDePalabras[0]}")
print(f"El indice 1 es: {arrayDePalabras[1]}\n\n")


print("Unimos las cadenas y las separamos ahora por ' - '")
nuevaCadena = " - ".join(arrayDePalabras)
print(f"La nueva cadena es: {nuevaCadena}") # Ahora la cadena es: Hola - Lexpin



system("cls")





# startswith() y endswith(): Verifica si una cadena comienza o termina con una subcadena específica.
texto = "Hola, ¿cómo estás?"

# Verificamos si el texto comienza con la secuencia de caracteres "Hola"
empieza_con_hola = texto.startswith("Hola")

# Verificamos si el texto termina con la secuencia de caracteres "?"
termina_con_signo = texto.endswith("?")

print(f"Es texto a evaluar es: {texto}")
print(f"¿Comienza con 'Hola' ?: {empieza_con_hola}")  # True
print(f"¿Termina con '?' ?: {termina_con_signo}")  # True

File: JavaScript/Seccion_2/Metodos_de_Strings.js
//Archivo 13

// Los strings o cadenas de caracteres cuentan con metodos para ser manipulados, estos metodos son funciones especiales
// que nos permiten realizar operaciones para modificar las cadenas de caracteres o strings.
// entre los principales metodos se encuentran:


let saludo = " Hola Mundo, esto es un mensaje de PRUEBA  ";


//  length: No es un metodo, es una propiedad que devuelve la longitud de una cadena. (No recibe parametros)
console.log( saludo.length );



//  toUpperCase() y toLowerCase(): convierten una cadena a mayúsculas o minúsculas, respectivamente. (No reciben parametros)
console.log(saludo.toLowerCase());
console.log(saludo.toUpperCase());



//  indexOf() y lastIndexOf(): buscan un valor específico en una cadena y devuelven su posición.
console.log(saludo.indexOf("e"));
console.log(saludo.lastIndexOf("e"));




//  slice(): extrae una porción de una cadena, Recibe dos parámetros (desde, hasta).
console.log(saludo.slice(16, 21));




//  replace(): reemplaza un valor específico en una cadena con otro valor.
console.log(saludo.replace("Mundo","Planeta"));




//  concat(): une dos o más cadenas.
console.log(saludo.concat(",Esto se acaba de concatenar"));




//  split(): divide una cadena en un array de subcadenas, su argumento sirve para delimitar la cadena original.
// Es decir el argumento indica donde se haran los cortes en la cadena.

const fragmentos = saludo.split(" ");
console.log(fragmentos);




// join(): permite unir los elementos de un array de caracteres en una nueva cadena, recibe como parametro
// el separador con el cual se uniran las partes.
const fragmentosUnidos = fragmentos.join(" - ");
console.log(fragmentosUnidos); 




//  trim(): elimina los espacios en blanco al principio y al final de una cadena. (No recibe parametros)

console.log(saludo);
console.log(saludo.trim());






//  startsWith() y endsWith(): determinan si una cadena comienza o termina 
//  con un sufijo(Secuencia al inicio) o prefijo(Secuencia al final) específico.

const mensaje = "Hola, te saludo cordialmente, mi nombre es Diego";

console.log(mensaje.startsWith("Hola"));
console.log(mensaje.endsWith("Diego"));





// Ejemplo usando varios metodos

const saludo2 = "  Hola Mundo esto es un mensaje de PRUEBA  ";
console.log(saludo2);


// Convertimos toda la cadena en mayúsculas
const saludo2Mayusculas = saludo2.toUpperCase();


// Quitamos los espacios al inicio y fin
const saludoSinEspacios = saludo2Mayusculas.trim();
console.log(saludoSinEspacios);


// Dividimos la cadena en un array, cortando donde existan espacios.
const fragmentosDeSaludo = saludoSinEspacios.split(" ");
console.log(fragmentosDeSaludo);


// Unimos los elementos del array en una nueva cadena agregando "lexpin" en medio de cada palabra.
const saludoConLExpin = fragmentosDeSaludo.join(" lexpin ");
console.log(saludoConLExpin);


/*
    Ejercicios para realizar en casa:

    1: Elabora un programa que te permita encriptar un mensaje con 16 caracteres de cifrado.


    2:  Elabora un programa que te permita determinar si un correo electronico es "gmail", "hotmail" o "outlook".
        Posteriormente se debe mostrar el dominio del correo electronico en un mensaje.
        Si el correo no termina con alguna de las 3 extensiones previamente mencionadas, se debe informar al usuario que
        utilice un correo admitido por el sistema.

    3:  Elabora un programa que mida la seguridad de una contraseña basandose solamente en la longitud de la cadena.
        Si tiene 2 caracteres, la seguridad es 1. 
        Si tiene 3 caracteres, la seguridad es 3.
        Si tiene 4 caracteres, la seguridad es 5.
        Si tiene 5 caracteres, la seguridad es 7.
        Si tiene 6 caracteres, la seguridad es 9.
        Si tiene mas de 6 caracteres, la seguridad es 10.
*/